Add FaunaDB Guide (#1178)
zeit/docs: pages/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now.mdx

import Guide from '~/components/layout/guide'
import Snippet from '~/components/snippet'
import { InlineCode } from '~/components/text/code'
import { Image } from '~/components/media'
import Caption from '~/components/text/caption'
import Note from '~/components/text/note'
import { GenericLink } from '~/components/text/link'
import Card from '~/components/card'
import { Code } from '~/components/code'

export const meta = {
  title: 'Create and Deploy a FaunaDB-Powered Node.js API with ZEIT Now',
  description:
    'Create a FaunaDB-Powered Node.js API and deploy it with ZEIT Now.',
  published: '2019-08-27T17:00:00.000Z',
  authors: ['danieltodonnell', 'msweeneydev'],
  url: '/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now',
  image: `${
    process.env.ASSETS
  }/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now/card.png`,
  editUrl:
    'pages/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now.mdx',
  lastEdited: '2019-08-27T20:17:12.000Z'
}

> [FaunaDB](https://fauna.com/) is a serverless cloud database that gives you ubiquitous, low latency access to app data, without sacrificing data correctness.

Reviewer: I know this is the FaunaDB brand, but I'm not quite sure `ubiquitous` is easy to understand for everyone, especially when we know we have readers that do not have English as a first language.

Maybe:
```suggestion
[FaunaDB](https://fauna.com/) is a serverless cloud database that gives you low latency access to app data around the world.
```

> This guide walks you through making a FaunaDB account, creating a sample database, and then connecting to it using a Serverless Function deployed with [ZEIT Now](/now).

Reviewer: Mentioning that I have to create an account first sounds long. Perhaps simplify this to the following:
```suggestion
This guide walks you through making a FaunaDB database and then connecting to it using a Serverless Function deployed with [ZEIT Now](/now).
```

Author: Agreed, especially as the guide mentions it assumes an existing account.

## Step 1: Connecting to FaunaDB

> To start this example, you need to have [created a FaunaDB account](https://dashboard.fauna.com/accounts/register). Once logged in to FaunaDB, create a child database named `zeit` _(names are case sensitive)_.

Reviewer: ```suggestion
To start, you need to have [created a FaunaDB account](https://dashboard.fauna.com/accounts/register). Once logged in to FaunaDB, create a child database named `zeit` _(names are case sensitive)_.
```

<Image
  src={`${
    process.env.ASSETS
  }/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now/fauna-create-zeit-db.png`}
  width={2048 / 2.5}
  height={1628 / 2.5}
  oversize
  shadow
/>
<Caption>Creating a new database in the FaunaDB dashboard.</Caption>

After clicking **Save**, you will be taken to the **Database Overview** page.

From here, the easiest way to create a demo schema with some data is to execute a _Fauna Query Language (FQL)_ script file.

> Select the **Shell** tab from the sidebar, copy in [this FQL script](https://raw.githubusercontent.com/danieltodonnell/faunadb-zeit-fql-demo/master/fql-scripts/ecommerce.fql) and click **Run Query**.

Reviewer: We can't reference UI prone to change. I'm also not sure about linking to the FQL script via `this FQL script` since some may look beyond that link. What are your thoughts there?

```suggestion
Select the **Shell** tab, then copy in [this FQL script](https://raw.githubusercontent.com/danieltodonnell/faunadb-zeit-fql-demo/master/fql-scripts/ecommerce.fql) and click **Run Query**.
```

Author: I am happy with using `this FQL script` as a link, I think it's easy to understand. How would you prefer to handle it?

<Image
  src={`${
    process.env.ASSETS
  }/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now/fauna-zeit-db-shell.png`}
  width={2048 / 2.5}
  height={1536 / 2.5}
  oversize
  shadow
/>

> <Caption>Adding data to a FaunaDB database from the dashboard using a FQL script.</Caption>

Reviewer: ```suggestion
<Caption>Adding data to a FaunaDB database from the dashboard using an FQL script.</Caption>
```

> After executing the script, the **Collections** page will look like this, you are now ready to create a connection to the database.

Reviewer: Try not to refer to images since people can use screen readers that don't show/give context through the image.

```suggestion
After executing the script, the **Collections** page will present you with dummy data. You are now ready to create a connection to the database.
```

<Image
  src={`${
    process.env.ASSETS
  }/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now/fauna-zeit-db-collections.png`}
  width={2048 / 2.5}
  height={1536 / 2.5}
  oversize
  shadow
/>
<Caption>The Collections page on the FaunaDB dashboard after adding data to the database.</Caption>

## Step 2: Create an Access Key

> Go to the [**Security** page](https://dashboard.fauna.com/keys/@db/zeit) and click the **New Key** button. Enter the following data in the fields:

Reviewer: ```suggestion
Go to the [**Security** page](https://dashboard.fauna.com/keys/@db/zeit) and create a new key. Enter the following data in the fields:
```

- Database: `zeit`
- Role: `Admin`
- Key Name: `access`
- Priority: `1`

> Once entered, click the **Save** button, which will take you to the **Keys** page.

Reviewer: ```suggestion
Save, once you have entered the required data. FaunaDB will then take you to the **Keys** page.
```

Author: I'm not sure about the order with this change. I think it will make more sense for readers to enter the data, then save.

Reviewer: Updated the suggestion @msweeneydev 🙏

On the **Keys** page, you will receive a unique hash called a **Secret**. The **Secret** is your method of securely identifying yourself to the database.

<Note warning>
  This is the only time you will see this hash, so record it somewhere private.

>   However, if the has is lost, a new one can be created.

Reviewer: ```suggestion
  However, if the hash is lost, a new one can be created.
```

</Note>

## Step 3: API Setup

> Get started by creating a project directory and `cd` into it:

Reviewer: ```suggestion
Get started creating your API by making a project directory and moving into it:
```

<Snippet dark text="mkdir faunadb-demo && cd faunadb-demo" />
<Caption>
  Creating and entering into the <InlineCode>/faunadb-demo</InlineCode>{' '}directory.
</Caption>

Next, [initialize](https://docs.npmjs.com/cli/init) the project:

<Snippet dark text="npm init -y" />
<Caption>
  Initializing the project, this creates a <InlineCode>package.json</InlineCode>{' '}file.
</Caption>

> Next, install [the FaunaDB Javascript client](https://github.com/fauna/faunadb-js), this helps you connect to and query the database:

Reviewer: ```suggestion
Continue to install [the FaunaDB JavaScript client](https://github.com/fauna/faunadb-js) which allows you connect to, and query, the database:
```

Author: 'which allows you connect to' => 'which allows you to connect to'?

Reviewer: Good catch!

<Snippet dark text="npm i faunadb" />
<Caption>
  Adding the <InlineCode>faunadb</InlineCode> Node.js client as a <GenericLink href="https://docs.npmjs.com/specifying-dependencies-and-devdependencies-in-a-package-json-file">dependency</GenericLink> to the project.
</Caption>

> To connect, you need the secret received earlier. To use the secret without hard-coding it into your project, add your connection string as a [Now Secret](https://zeit.co/docs/v2/deployments/environment-variables-and-secrets/#securing-environment-variables-using-secrets) using [Now CLI](https://zeit.co/download#now-cli) to keep it secure:

Reviewer: ```suggestion
To connect, you need to use the secret obtained in [Step 2](#step-2:-create-an-access-key). To use the secret without hard-coding it into your project, add your connection string as a [Now Secret](https://zeit.co/docs/v2/deployments/environment-variables-and-secrets/#securing-environment-variables-using-secrets) using [Now CLI](https://zeit.co/download#now-cli) to keep it secure:
```

> <Snippet dark text="now secrets add faunadb_secret_key YourFaunaDBSecret" />

Reviewer: ```suggestion
<Snippet dark text="now secrets add faunadb_secret_key [Your FaunaDB Secret]" />
```

> <Snippet dark text="now secrets add faunadb_secret_key [Your FaunaDB Secret]" />

Reviewer 2: I was going through the guide and I didn't have `now` installed in my machine at this point. I've noticed that there's a message pointing to a page for installing the CLI later on in [Step 5: Deploying](https://github.com/zeit/docs/blob/temp-fauna-merge/pages/guides/deploying-a-faunadb-powered-api-with-node-and-zeit-now.mdx#step-5-deploying):

> If you have not yet installed Now, you can do so by installing Now CLI.

Maybe that sentence could be moved up here in the form of an info/note box?

Author: Thank you for your feedback here @lregnier, we use this pattern in all our guides currently. That being said, it's something we have discussed recently and may look at updating for all soon.

<Caption>
  Adding an environment variable to the project as a <GenericLink href="/docs/v2/serverless-functions/env-and-secrets/#adding-secrets">Now Secret</GenericLink>.
</Caption>

> Now, create a [`now.json` file](https://zeit.co/docs/v2/deployments/configuration/) that tells Now to make available the [Secret](https://zeit.co/docs/v2/deployments/environment-variables-and-secrets/#securing-environment-variables-using-secrets) that you defined:

Reviewer: ```suggestion
Create a [`now.json` file](https://zeit.co/docs/v2/deployments/configuration/) that tells Now to make available the [Secret](https://zeit.co/docs/v2/deployments/environment-variables-and-secrets/#securing-environment-variables-using-secrets) that you defined, to your project:
```

```json
{
  "env": {
    "FAUNADB_SECRET_KEY": "@faunadb_secret_key"
  }
}
```

<Caption>
  An example <InlineCode>now.json</InlineCode> file that provides the app with
  an environment variable.
</Caption>

<Note hint>
  To use this environment variable{' '}
  <GenericLink href="/docs/v2/serverless-functions/introduction/#local-development">
    during local development
  </GenericLink>
  , you need to add it to a<InlineCode>.env</InlineCode> file.
</Note>

## Step 4: Writing the Serverless Function

> It's time to create the Node.js API endpoint that will fetch the data from FaunaDB. Create an `/api` directory with an `index.js` file inside:

Reviewer: ```suggestion
Create the Node.js API endpoint that will fetch the data from FaunaDB by making an `/api` directory with an `index.js` file inside:
```

> <Snippet dark text="mkdir -p api/index.js && cd api" />

Reviewer 2: This command seems to create a folder name `index.js`, instead of a file. Something like below should create a file within the new folder:

```
mkdir api && touch api/index.js && cd api
```

Author: Good spot! I've actually removed this command. The reason being the `touch` command can be an issue for Windows users that aren't using the Linux Subsystem. Where files need to be created, we normally state that in text without accompaniment from a <Snippet>.

</Caption>

The `index.js` file will act as the default endpoint for getting information from your database. The file should contain the following code:

```js
const faunadb = require('faunadb')

// your secret hash
const secret = process.env.FAUNADB_SECRET_KEY
const q = faunadb.query
const client = new faunadb.Client({ secret })

module.exports = async (req, res) => {
  try {
    const dbs = await client.query(
      q.Map(
        // iterate each item in result
        q.Paginate(
          // make paginatable
          q.Match(
            // query index
            q.Index('all_customers') // specify source
          )
        ),
        ref => q.Get(ref) // lookup each result by its reference
      )
    )
    // ok
    res.status(200).json(dbs.data)
  } catch (e) {
    // something went wrong
    res.status(500).json({ error: e.message })
  }
}
```

<Caption>
  An example <InlineCode>/api/index.js</InlineCode> file that retrieves
  information from the database.
</Caption>

## Step 5: Deploying

With the project complete, you are ready to deploy it with [ZEIT Now](/now).

If you have not yet installed Now, you can do so by [installing Now CLI](/download).

> You can now deploy the project with **just a single command**:

Reviewer: ```suggestion
You can now deploy the project with **a single command**:
```

> <Snippet dark text="now" />

Reviewer 2: It might be worthy to specify here that the command `now` should be run in the root folder of the project. If you are just copy-pasting or typing the commands as they appear in the guide, at this point you will be within the `api` folder (per command in previous step). If running `now` within the `api` folder, it will be deployed the `index.js` file as a static resource. This will lead to an unexpected result, which might be difficult to new comers to understand and resolve.

Author: As a result of the previous change (no longer moving into the `api` directory), this is no longer a problem. Thank you for being thorough here @lregnier.

<Caption>
  Deploying the project with the <InlineCode>now</InlineCode> command.
</Caption>

> You will shortly be given a deployment URL. With this, you can now access the contents of the FaunaDB database at the `/api` endpoint.

Reviewer: ```suggestion
You will be given a deployment URL, with which you can access the contents of the FaunaDB database at the `/api` endpoint.
```
